### PR TITLE
fix: soft or hard delete insight after soft or hard dashboard deletion

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

Dashboards and insights weren't clearing their relations on soft delete. So if you deleted either a dashboard or an insight after linking them, you couldn't then delete the other half of the pair.

Hard deletions cascaded to delete the relations

before-delete

## Changes

On soft delete of either insight or dashboard cascade to delete the `DashboardTile`

after delete

## How did you test this code?

unit tests and running locally

### CODE REVIEW

File: posthog/api/test/test_dashboard.py [hunk 1]
```diff
@@ -659,6 +659,116 @@ def test_patch_api_as_form_data(self):
         self.assertEqual(response.status_code, 200, response.content)
         self.assertEqual(response.json()["name"], "replaced")
 
+    def test_can_hard_delete_insight_after_hard_deleting_dashboard(self) -> None:
```
Comment (timgl) [inline posthog/api/test/test_dashboard.py:662]: flyby comment. Why do we allow both hard and soft deletes? We should only do soft deletes really so people can undo (once we get around to implementing that)

Comment (pauldambra) [reply posthog/api/test/test_dashboard.py:662]: I guess they come for free with DRF and we've not removed them

Comment (timgl) [reply posthog/api/test/test_dashboard.py:662]: I don't think there's any point in having both and testing both. Let's only allow soft deletes and remove the ability to hard delete.

**Comment by pauldambra** [comment]
Added forbid delete as a mixin so we can do this more widely.

But didn't change every model in this PR to keep it focussed on the bug fix

I'll followup in a separate PR to do the same for each other model with a deleted property

* cohort
* annotation
* action
* feature flag

**Review by benjackwhite — APPROVED** [review]

**Pull request merged**